dbms/src/DataStreams/ParallelInputsProcessor.h

```diff
@@ -205,27 +217,52 @@ class ParallelInputsProcessor
 
         try
         {
-            while (!finish)
-            {
-                InputData unprepared_input;
-                {
-                    std::lock_guard lock(unprepared_inputs_mutex);
+            loop(thread_num);
+        }
+        catch (...)
+        {
+            exception = std::current_exception();
+        }
 
-                    if (unprepared_inputs.empty())
-                        break;
+        if (exception)
+        {
+            handler.onException(exception, thread_num);
+        }
 
-                    unprepared_input = unprepared_inputs.front();
-                    unprepared_inputs.pop();
-                }
+        handler.onFinishThread(thread_num);
 
-                unprepared_input.in->readPrefix();
+        if (additional_inputs_at_end.empty())
+        {
+            if (0 == --active_threads)
+            {
+                handler.onFinish();
+            }
+        }
 
-                {
-                    std::lock_guard lock(available_inputs_mutex);
-                    available_inputs.push(unprepared_input);
-                }
+        {
```

Comment:
Here contains two steps:
1. all threads wait for a barrier.
2. init streams into a pair of `unprepared_inputs` and `available_inputs`.

My suggestions:
1. explicitly express the barrier semantic.
2. how about encapsulate streams, `unprepared_inputs`, and `available_inputs` into one struct named as `WorkingStreams`? It will have two instances, one for inputs, another for additional_inputs_at_end. Then you can first call `loop(thread_num, inputsStreams)` then call `loop(thread_num, additionalStreams)` and needn't to do initialization between them.